storage/storage_image.go

```diff
@@ -568,7 +568,35 @@ func (s *storageImageDestination) Commit(ctx context.Context) error {
 			if err2 != nil {
 				return errors.Wrapf(err2, "error reading layer %q for blob %q", layer, blob.Digest)
 			}
-			defer diff.Close()
+			// Copy the layer diff to a file.  Diff() takes a lock that it holds
+			// until the ReadCloser that it returns is closed, and PutLayer() wants
+			// the same lock, so the diff can't just be directly streamed from one
+			// to the other.
+			filename := filepath.Join(s.directory, fmt.Sprintf("%d", atomic.AddInt32(&s.nextTempFileID, 1)))
```

Comment:
(Absolutely non-blocking: consider adding a helper function so that there is only one place managing the `nextTempFileID` atomic value.)

Reply:
Sounds simple enough.  Adding a `computeNextBlobCacheFile()` method.